src/playlist-selectors.js

```diff
@@ -103,91 +125,118 @@ export const comparePlaylistResolution = function(left, right) {
-  let resolutionPlusOneList = [];
-  let resolutionPlusOneSmallest = [];
-  let resolutionBestVariantList = [];
 
-  stableSort(sortedPlaylists, comparePlaylistBandwidth);
+  // convert the playlists to an intermediary representation to make comparisons easier
+  // and prevent us from re-determining bandwidth each time
+  let sortedPlaylistReps = sortedPlaylists.map((playlist) => {
+    let playlistWidth;
+    let playlistHeight;
+
+    if (playlist.attributes && playlist.attributes.RESOLUTION) {
+      playlistWidth = playlist.attributes.RESOLUTION.width;
+      playlistHeight = playlist.attributes.RESOLUTION.height;
+    }
+
+    return {
+      bandwidth: getBandwidth(playlist, useAverageBandwidth) || window.Number.MAX_VALUE,
+      width: playlistWidth,
+      height: playlistHeight,
+      playlist
+    };
+  });
+
+  stableSort(sortedPlaylistReps, (left, right) => left.bandwidth - right.bandwidth);
 
   // filter out any playlists that have been excluded due to
   // incompatible configurations or playback errors
-  sortedPlaylists = sortedPlaylists.filter(Playlist.isEnabled);
+  sortedPlaylistReps = sortedPlaylistReps.filter(
+    (rep) => Playlist.isEnabled(rep.playlist));
+
   // filter out any variant that has greater effective bitrate
   // than the current estimated bandwidth
-  bandwidthPlaylists = sortedPlaylists.filter(function(elem) {
-    return elem.attributes &&
-           elem.attributes.BANDWIDTH &&
-           elem.attributes.BANDWIDTH * Config.BANDWIDTH_VARIANCE < bandwidth;
-  });
+  let bandwidthPlaylistReps = sortedPlaylistReps.filter(
+    (rep) => rep.bandwidth && rep.bandwidth * Config.BANDWIDTH_VARIANCE < bandwidth
+  );
+
+  let highestRemainingBandwidthRep =
+    bandwidthPlaylistReps[bandwidthPlaylistReps.length - 1];
 
   // get all of the renditions with the same (highest) bandwidth
   // and then taking the very first element
-  bandwidthBestVariant = bandwidthPlaylists.filter(function(elem) {
-    return elem.attributes.BANDWIDTH === bandwidthPlaylists[bandwidthPlaylists.length - 1].attributes.BANDWIDTH;
-  })[0];
+  let bandwidthBestRep = bandwidthPlaylistReps.filter(
+    (rep) => rep.bandwidth === highestRemainingBandwidthRep.bandwidth
+  )[0];
 
   // sort variants by resolution
-  stableSort(bandwidthPlaylists, comparePlaylistResolution);
+  stableSort(bandwidthPlaylistReps, (left, right) => {
+    const leftWidth = left.width || window.Number.MAX_VALUE;
+    const rightWidth = right.width || window.Number.MAX_VALUE;
 
-  // filter out playlists without resolution information
-  haveResolution = bandwidthPlaylists.filter(function(elem) {
-    return elem.attributes &&
-           elem.attributes.RESOLUTION &&
-           elem.attributes.RESOLUTION.width &&
-           elem.attributes.RESOLUTION.height;
+    return leftWidth - rightWidth;
   });
 
+  // filter out playlists without resolution information
+  let haveResolution = bandwidthPlaylistReps.filter((rep) => rep.width && rep.height);
```

Comment:
If you do this before the sort above, and then stable sort `haveResolution` you can remove the local constants and make it a one-line like the bandwidth sort